src/main/java/mekanism/common/lib/inventory/personalstorage/PersonalStorageManager.java

```diff
@@ -0,0 +1,95 @@
+package mekanism.common.lib.inventory.personalstorage;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import mekanism.api.DataHandlerUtils;
+import mekanism.api.NBTConstants;
+import mekanism.api.annotations.ParametersAreNotNullByDefault;
+import mekanism.common.lib.MekanismSavedData;
+import mekanism.common.util.ItemDataUtils;
+import mekanism.common.util.SecurityUtils;
+import net.minecraft.MethodsReturnNonnullByDefault;
+import net.minecraft.nbt.ListTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.fml.loading.FMLEnvironment;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+@MethodsReturnNonnullByDefault
+@ParametersAreNotNullByDefault
+public class PersonalStorageManager {
+    private static final Map<UUID, PersonalStorageData> STORAGE_BY_PLAYER_UUID = new HashMap<>();
+
+    private static PersonalStorageData forOwner(UUID playerUUID) {
+        return STORAGE_BY_PLAYER_UUID.computeIfAbsent(playerUUID, uuid->MekanismSavedData.createSavedData(PersonalStorageData::new, "personal_storage" + File.separator + uuid));
+    }
+
+    /**
+     * Only call on the server. Gets or creates an inventory for the supplied stack
+     *
+     * @param stack Personal storage ItemStack (type not checked) - will be modified if it didn't have an inventory id
+     * @return the existing or new inventory
+     */
+    public static PersonalStorageItemInventory getInventoryFor(ItemStack stack) {
+        UUID owner = SecurityUtils.get().getOwnerUUID(stack);
+        if (owner == null) {
+            throw new IllegalStateException("Stack has no owner!");
+        }
+        UUID invId = getInventoryId(stack);
+        PersonalStorageItemInventory storageItemInventory = forOwner(owner).getOrAddInventory(invId);
+        ListTag legacyData = ItemDataUtils.getList(stack, NBTConstants.ITEMS);
```

Comment:
Do we want a TODO to remove this at some point? Such as `//TODO - 1.21: Remove legacy loading`?